### ISSUE TITLE
ofAVFoundationPlayer can't dance backwards

### ISSUE THREAD

**Issue by kylemcdonald** [openframeworks/openFrameworks#3629]
(Accidentally submitted before I filled out the text.)

The brand new ofAVFoundationPlayer can step one frame back at a time, but can't handle a negative play rate. Instead it stays on the current frame, or toggles back and forth between the previous and current frame. Haven't looked at possible causes yet.

To reproduce, build videoPlayerExample and move the mouse left of center.

**Comment by julapy** [comment]
ah ok, ill investigate.

**Comment by admsyn** [comment]
Did a quick bit of research on this, as far as I can tell as-is ofAVFoundationPlayer is struck with whatever AVFoundation reports with regards to backwards playback capabilities. 

When loading an asset, you get 3 read-only bool properties back: [canPlayReverse](https://developer.apple.com/library/ios/documentation/AVFoundation/Reference/AVPlayerItem_Class/index.html#//apple_ref/occ/instp/AVPlayerItem/canPlayReverse), [canPlaySlowReverse](https://developer.apple.com/library/ios/documentation/AVFoundation/Reference/AVPlayerItem_Class/index.html#//apple_ref/occ/instp/AVPlayerItem/canPlaySlowReverse), and [canPlayFastReverse](https://developer.apple.com/library/ios/documentation/AVFoundation/Reference/AVPlayerItem_Class/index.html#//apple_ref/occ/instp/AVPlayerItem/canPlayFastReverse) corresponding to rates of -1, 0 to -1 and less than -1 respectively (presumably codec / hardware dependent). The fingers vid gives me false for all of these on a MacBook Pro, 10.10.

I doubt it's a case of Cannot Be Done but it may be that a straight-up AVPlayer isn't going to be able to do it.

**Comment by urshofer** [comment]
I've posted the same question to the forum [http://forum.openframeworks.cc/t/avfoundationplayer-and-negative-speed/18684] - It's probably also dependent on the media, sometimes it works with i-frame only formats like prores…

**Comment by i-n-g-o** [comment]
ofAVFoundationPlayer uses AVAssetReader to get the audio- and videodata.
An AVAssetReader is not capable of reading an asset backwards.

after playing around with the implementation i have backwards playing running, but it is really slow... doing a seekToTime for every frame.
(reducing the timeRange to one frame, so the reader would not unnecessarily preload frames)

in order to have this running faster it would be better to jump back to the last keyframe in the video, let the assetReader preload the frames, read the frames off the AssetReader, cache the frames and put it out in reverse order. while getting the frames, the next previous junk should get prepared, preloaded and cached...

anybody knows how to get the keyframes of an asset with AVFoundation?
if the video file would only consist of keyframes, then it should still read a couple of frames and cache them.

maybe then it would be more consistent to use the same technique for forward palying as well (chaching files between keyframes). the advantage beeing that changing from forward to backward playing would be immediate since the frames are already cached.

... or we use a AVPlayerLayer together with a AVPlayer and manage to get the texture out of it.

**Comment by admsyn** [comment]
In case anyone's plugging away at this, I had a minor revelation in an unrelated project. The restrictive playback rate options seem related to the types of tracks enabled in the file, and specifically audio tracks. Through a bit of KVO you can observe the "tracks" property on the AVPlayerItem and disable the audio tracks when loaded. 

In my case this was the difference between a playback rate locked to 1 and a playback rate of "pretty much whatever", included backwards playback.

Code looks like this:
```objective-c
- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
	if(object == _playerItem && [keyPath isEqualToString:@"tracks"]) {
		for(AVPlayerItemTrack * track in _playerItem.tracks) {
			if([track.assetTrack.mediaType isEqualToString:AVMediaTypeAudio]) {
				track.enabled = NO;
			}
		}
	}
}
```

**Comment by i-n-g-o** [comment]
In the project you mention, do you use an AVAssetReader to get access to the data? or do you use an AVPlayer (with an AVPlayerItem) to playback media?

**Comment by admsyn** [comment]
Just a regular AVPlayer (with an AVPlayerLayer & AVPlayerItem).

**Comment by bilderbuchi** [comment]
@openframeworks/video @openframeworks/macos a quick update on the status of this, please? is this something we will fix for 0.9.0?

**Comment by pizthewiz** [comment]
I think @admsyn determined that the oF AV Foundation player _is_ capable of reverse playback, however not all media types (specifically audio tracks) allow it. 

Adding the iOS keyword since the player is now shared.

**Comment by i-n-g-o** [comment]
depending on compression format AV Foundation Player itself is capable of playing media backwards.

ofAVFoundationPlayer does use AVAssetReader to read media frames. To my knowledge AVAssetReader is _not_ capable of playing back media backwards (preloading frames in reverse order...)

please see my comment above.

for reverse playback ofAVFoundationPlayer has to work differently, caching frames in the past.
this needs a bit of work.

one crucial detail is to get to know the i-frame rate of the media. i did not find anything about this.
if anybody knows how to obtain the i-frame rate of a movie with ACFoundation/CoreMedia it would be very helpful and could be the start of reconstructing the player.

how important is reverse palyback for 0.9.0?

**Comment by bilderbuchi** [comment]
> how important is reverse palyback for 0.9.0?

while it is a regression, I think it's not very important. I'll push this back to 0.9.1

**Comment by urshofer** [comment]
i think - from a creative coding viewpoint - backwards, slow motion, frame by frame is kind of important. and it would be great to keep the functionalities between the platforms in sync. a different approach would be to use gstreamer as video framework as widely as possible. it's probably not the fastest, but the most open and versatile system and fit probably best what oF is meant to be.


Am 17.06.2015 um 20:31 schrieb Christoph Buchner <notifications@github.com>:

**Comment by kylemcdonald** [comment]
in my opinion the absolute minimum thing we need to do here before 0.9.0 is to log an `ofWarning` about negative play rates not being support for AVFoundationPlayer, and convert it to a 0 play rate. otherwise the jittery behavior will create confusion when someone runs into this.

**Comment by i-n-g-o** [comment]
i had another look into AVFoundation and found a solution for backwards playback if we use minimum OSX 10.8.

would be cool if somebody could test this. (also on iOS)
https://github.com/i-n-g-o/openFrameworks/tree/AVFoundationPlayer_dance_backwards

some details:
from OSX 10.8 on we have AVPlayerItemVideoOutput which can be added as an output to a AVPlayerItem. AVPlayerItemVideoOutput always is sync with the playerItems current time. as long as the AVPlayerItem is able to play the asset reverse, all is good.

When i tested with AVPlayerItemVideoOutput only (skipping the AVAssetReader stuff) i figured some issues when playing back 1080p25 footage. not even mentioning about 4k...
using the AVAssetReader constructions preloads the frames and playback is much smoother.

ofAVFoundationVideoPlayer now is using a mix of both. playback forward is done by pulling frames out of the AVAsserReader. playing backwards uses AVPlayerItemVideoOutput if supported and drops errors if AVPlayerItemVideoOutput is not available.

**Comment by i-n-g-o** [comment]
check out the PR:
https://github.com/openframeworks/openFrameworks/pull/4094

**Comment by bilderbuchi** [comment]
Closed by #4094
